### DISCUSSION TITLE
Consider "redirects as pages" notes from Ethan Jucovy

### DISCUSSION THREAD

**Discussion by davecranwell** [wagtail/wagtail#11383] [Ideas]
http://www.coactivate.org/projects/ejucovy/blog/2014/05/10/wagtail-notes-managing-redirects-as-pages/

Displaying redirects in their position in the Explorer tree seems like a pretty good idea for starters. Other notes may also be worth investigating.

**Comment by zerolab** [comment]
FWIW, I've implemented something based on Ethan's notes on a project and that proved very helpful.

One of our use cases: needed to add a link in the menu to a specific anchor tag (a UI tab) 

D

**Comment by alej0varas** [comment]
I've created a gist borrowing Ethan's code and using `LinkFields` to have redirects to pages, uls and documents. It solves my need of a link to an external url placed in the top menu but also work for secondary menu.

https://gist.github.com/alej0varas/e7e334643ceab6e65744

**Comment by gasman** [comment]
Some assorted thoughts on this, as the subject of redirects has come up a couple of times today...

In many ways, the redirect-as-page approach is more elegant than wagtailredirects: they're more visible in the admin, have fewer moving parts (no need for middleware or a separate admin module) and can be used in situations where the source URL isn't a 404 (see #2533).

However, there are a couple of downsides that make it not a good fit for some scenarios:
- redirecting from `/a/deeply/nested/path/with/lots/of/slashes` requires setting up a bunch of nested pages
- you can't set up a redirect from a URL that contains non-slug-safe characters
- it doesn't have the special case handling for matching an exact query string first (`/myblog/show_article?id=123`) then falling back on a non-query-string match. You could make the `RedirectPage` model smarter to allow redirecting to different places based on querystring, but that's probably going to be much less intuitive than just treating the entire URL as a string match
- they're governed by the page permission model, and so you can't treat them as a different kind of "thing" for permission management purposes, e.g. specify that only superusers are allowed to create redirects. But arguably the page permission model can be a good thing too - it means you can say "this user can create redirects within `/departments/engineering/`.

All in all I think there's room for both concepts to exist in Wagtail (perhaps as two distinct contrib modules?) - although then we have the question of how to communicate that to users without looking like we've implemented the same thing twice in two incompatible ways. It strikes me that wagtailredirects is a better fit for preserving legacy URLs when migrating away from an old site, whereas redirects-as-pages is a better fit for shortcuts / directing a user around the site - so perhaps there's some way of naming them to represent those two roles.
